plaidml/bridge/openvino/ops/detection_output.cpp

```diff
@@ -0,0 +1,223 @@
+    auto priors_shape_variance = Priors.compute_shape().sizes()[1];
+
+    std::vector<int64_t> location_shape = {batch, num_priors, 4};
+    std::vector<int64_t> confidence_shape = {batch, num_priors, num_classes};
+    std::vector<int64_t> priors_shape = {batch, priors_shape_variance, num_priors, prior_size};
+
+    edsl::Tensor location = edsl::reshape(Location, location_shape);
+    edsl::Tensor confidence = edsl::reshape(Confidence, confidence_shape);
+    edsl::Tensor priors = edsl::reshape(Priors, priors_shape);
+
+    edsl::Tensor variances;
+    if (variance_encoded_in_target) {
+      // Fill variances with 1s.
+      priors =
+          op::slice(priors).add_dim(0, batch).add_dim(0, 1).add_dim(0, num_priors).add_dim(prior_offset, prior_size);
+      auto one = cast(edsl::Tensor{1}, priors.dtype());
+      auto var_shape = priors.compute_shape().sizes();
+      variances = op::broadcast(one, var_shape, {});
+    } else {
+      variances = op::slice(priors).add_dim(0, batch).add_dim(1, 2).add_dim(0, num_priors).add_dim(0, prior_size);
+      // Reshape the variances tensor from {batch, 1, num_priors, 5} to {batch, 1, num_priors, 4};
+      variances = edsl::reshape(variances, {batch, 1, num_priors * prior_size});
+      // Truncate the variances tensor.
+      variances = op::slice(variances).add_dim(0, batch).add_dim(0, 1).add_dim(0, num_priors * 4);
+      variances = edsl::reshape(variances, {batch, 1, num_priors, 4});
+      priors =
+          op::slice(priors).add_dim(0, batch).add_dim(0, 1).add_dim(0, num_priors).add_dim(prior_offset, prior_size);
+    }
+
+    variances = op::squeeze(variances, {1});
+    priors = op::squeeze(priors, {1});
+
+    edsl::Tensor decoded_bboxes;
+    if (with_add_pred) {
+      // Update confidence if there are 5 inputs.
+      Tensor IX = edsl::index({edsl::TensorDim(num_priors)}, 0);
+      IX = IX * 2 + 1;
+      edsl::Tensor arm_conf = edsl::gather(ArmConfidence, IX).axis(1);
+      arm_conf = op::repeat(op::unsqueeze(arm_conf, {-1})).count(num_classes).axis(2);
+      confidence = edsl::select(arm_conf < objectness_score, cast(edsl::Tensor{0.0f}, DType::FLOAT32), confidence);
+
+      // Decode bouding boxes
+      edsl::Tensor arm_loc = edsl::reshape(ArmLocation, location_shape);
+      auto decoded_priors =
+          DecodeBoxes(priors, variances, arm_loc, code_type, i_h, i_w, batch, num_priors, clip_before_nms);
+      decoded_bboxes =
+          DecodeBoxes(decoded_priors, variances, location, code_type, i_h, i_w, batch, num_priors, clip_before_nms);
+    } else {
+      decoded_bboxes =
+          DecodeBoxes(priors, variances, location, code_type, i_h, i_w, batch, num_priors, clip_before_nms);
+    }
+
+    // Transpose the confidence to match the input shape of `scores` in NMS.
+    // confidence -> {batch, num_boxes, num_classes}
+    // scores -> {batch, num_classes, num_boxes}
+    edsl::Tensor nms_conf = op::transpose(confidence, edsl::make_tuple<int64_t>({0, 2, 1}));
```

Comment:
If we change axis in NMS to do comparison, can we avoid this transpose?

Reply:
Yes but by doing so NMS will be different as the specs described. I'm not sure which way is better.

Reply:
As we may add new mode to NMS which will finally change the codes, i think we can add changes in that mode

Reply:
Sure. I will look into that.